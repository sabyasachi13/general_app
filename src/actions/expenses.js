import  uuid from 'uuid';
import database from '../firebase/firebase';
export const addExpense =(expense)=>({
    type:"ADD_EXPENSE",
    expense
});

export const startAddExpense =(expenseData={})=>{
    return(disapatch,getState)=>{
        const uid =getState().auth.uid;
        const {
            description="",
            note="",
            amount=0,
            createdAt=0
        }=expenseData;
        const expense ={description,note,amount,createdAt};

        return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
            disapatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
    }
};

export const removeExpense =({id}={})=>({
    type:"REMOVE_EXPENSE",
    id
    
});

export const editExpense =(id,updates)=>({
    type:'EDIT_EXPENSE',
    id,
    updates
});
//database.ref().update({
    //    stressLevel : 9,
    //    "job/company" : "amazon",
    //    "location/city": "seattle"
    // });
    
export const startEditExpense=(id,updates)=>{
    return (dispatch,getState)=>{
        const uid =getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update({
            ...updates
        }).then(()=>{
            console.log(updates);
            dispatch(editExpense(id,updates));
        })
    }
}

export const setExpenses=(expenses)=>({
    type:"SET_EXPENSES",
    expenses
});


export const startSetExpenses =()=>{
    return(dispatch,getState)=>{ 
        const uid =getState().auth.uid;
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot)=>{
            const expenses=[];
            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id:childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses))
        })
    }
}

export const startRemoveExpense=({id}={})=>{
    return (dispatch,getState)=>{
        const uid =getState().auth.uid;
       return  database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}));
        })
    }
}