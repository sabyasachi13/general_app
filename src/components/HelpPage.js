import React from 'react';
const helpPage =()=>(
    <div>
        This is help expense page
    </div>
);

export default helpPage;