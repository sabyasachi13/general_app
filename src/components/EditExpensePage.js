import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense,removeExpense} from '../actions/expenses';
//class based component
//setup mapDispatchToProps
export class EditExpensePage extends React.Component{
    onSubmit=(expense)=>{
        this.props.editExpense(this.props.expense.id,expense)
        this.props.history.push('/');
    }
    onClick=()=>{
        this.props.removeExpense({id:this.props.expense.id});
        this.props.history.push('/');
    }
    render(){
        return(
            <div>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                    expense = {this.props.expense}
                />
                <button onClick={this.onClick}>Remove</button> 
            </div> 
        )
    }
}

const mapDispatchToProps=(dispatch,props)=>{
    return{
        editExpense:(id,expense) => dispatch(editExpense(id,expense)),
        removeExpense:(id)=>dispatch(removeExpense(id))
    }
}
const mapStateToProps=(state,props)=>{
    return{
        expense :state.expenses.find((expense)=>{
            return expense.id ===props.match.params.id
        })
    }
}
export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);