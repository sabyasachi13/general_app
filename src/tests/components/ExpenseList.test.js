import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test ('Should render Expense List With expense',()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);

    expect(wrapper).toMatchSnapshot();
});

test('should render when Expense list is empty ',()=>{
    const wrapper = shallow(<ExpenseList expenses={[]}/>);

    expect(wrapper).toMatchSnapshot();
})