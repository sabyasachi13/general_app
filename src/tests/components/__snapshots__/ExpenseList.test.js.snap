// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Expense List With expense 1`] = `
<div>
  <ExpenseListItem
    amount={100}
    createdAt={0}
    description="gum"
    id="1"
    key="1"
    notes=""
  />
  <ExpenseListItem
    amount={12345}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    notes=""
  />
  <ExpenseListItem
    amount={12345}
    createdAt={345600000}
    description="credit"
    id="3"
    key="3"
    notes=""
  />
</div>
`;

exports[`should render when Expense list is empty  1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;
